src/sources/channel.rs

```diff
@@ -85,19 +84,26 @@ impl<T> SyncSender<T> {
     /// deliver the message. Readiness is signalled to the event loop
     /// *before* blocking.
     pub fn send(&self, t: T) -> Result<(), mpsc::SendError<T>> {
-        self.ping.ping();
-        self.sender.send(t)
+        let ret = self.try_send(t);
+        match ret {
+            Ok(()) => Ok(()),
+            Err(mpsc::TrySendError::Full(t)) => self.sender.send(t),
```

Comment:
It seems like you don't ping after this `send`?

Reply:
The previous call to `self.try_send()` will have pinged already.

Reply:
Yes, but there's no guarantee that it's `try_send` and `send` are done atomically? So you can ping() -> poll() -> send() -> lost wake up.

Reply:
You mean, if the event loops wakes up and empties the full channel before the sending thread gets the occasion to send its message?

Reply:
I suspect the post-ping would mostly cause spurious wakeups of the event loop overall, but I guess we can err on the side of safety. Spurious wakeups are better than delayed events.